File: pricing-core/src/test/kotlin/com/heenotane/pricing/PricingCalculatorTest.kt
// src/test/kotlin/com/heenotane/pricing/PricingCalculatorTest.kt
//
// PricingCalculator の基本テスト。
//
// 関連:
//   - PricingCalculator.kt
//   - Models.kt

package com.heenotane.pricing

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.math.BigDecimal
import java.util.stream.Stream

class PricingCalculatorTest {

    // ===================================================================
    // 税計算
    // ===================================================================

    @Test
    fun `tax excluded basic 398 yen 10pct`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("398"), taxIncluded = false, taxRate = BigDecimal("0.10"),
        )
        val bd = calculatePrice(offer)
        assertEquals(BigDecimal("398"), bd.preTaxPrice)
        assertEquals(BigDecimal("40"), bd.taxAmount)
        assertEquals(BigDecimal("438"), bd.basePriceInclTax)
    }

    @Test
    fun `tax included basic 398 yen 10pct`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("398"), taxIncluded = true, taxRate = BigDecimal("0.10"),
        )
        val bd = calculatePrice(offer)
        assertEquals(BigDecimal("362"), bd.preTaxPrice)
        assertEquals(BigDecimal("36"), bd.taxAmount)
        assertEquals(BigDecimal("398"), bd.basePriceInclTax)
    }

    @Test
    fun `zero tax rate`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("100"), taxIncluded = false, taxRate = BigDecimal.ZERO,
        )
        val bd = calculatePrice(offer)
        assertEquals(BigDecimal("0"), bd.taxAmount)
        assertEquals(BigDecimal("100"), bd.basePriceInclTax)
    }

    // ===================================================================
    // 割引
    // ===================================================================

    @Test
    fun `10 percent discount`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("1000"), taxIncluded = true, taxRate = BigDecimal("0.10"),
            percentageDiscount = BigDecimal("0.10"),
        )
        val bd = calculatePrice(offer)
        assertEquals(BigDecimal("100"), bd.discountAmount)
        assertEquals(BigDecimal("900"), bd.payableNow)
    }

    // ===================================================================
    // クーポン
    // ===================================================================

    @Test
    fun `coupon 30 yen`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("398"), taxIncluded = true, taxRate = BigDecimal("0.10"),
            couponDiscount = BigDecimal("30"),
        )
        val bd = calculatePrice(offer)
        assertEquals(BigDecimal("30"), bd.couponDiscount)
        assertEquals(BigDecimal("368"), bd.payableNow)
    }

    // ===================================================================
    // 送料
    // ===================================================================

    @Test
    fun `shipping fee added`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("330"), taxIncluded = true, taxRate = BigDecimal("0.10"),
        )
        val ctx = PurchaseContext(shippingFee = BigDecimal("80"))
        val bd = calculatePrice(offer, context = ctx)
        assertEquals(BigDecimal("80"), bd.shippingFee)
        assertEquals(BigDecimal("410"), bd.payableNow)
    }

    // ===================================================================
    // ポイント
    // ===================================================================

    @Test
    fun `point rate 1pct`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("398"), taxIncluded = true, taxRate = BigDecimal("0.10"),
            couponDiscount = BigDecimal("30"), pointRate = BigDecimal("0.01"),
        )
        val bd = calculatePrice(offer)
        assertEquals(BigDecimal("3"), bd.earnedPoints)
    }

    @Test
    fun `points missing warning`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("1000"), taxIncluded = true, taxRate = BigDecimal("0.10"),
        )
        val bd = calculatePrice(offer)
        assertTrue(bd.warnings.contains(WarningCode.POINTS_MISSING.value))
    }

    // ===================================================================
    // 容量換算
    // ===================================================================

    @Test
    fun `unit cost 500ml x2`() {
        val offer = Offer(
            id = "a", productName = "a", storeName = "a",
            price = BigDecimal("500"), taxIncluded = true, taxRate = BigDecimal("0.10"),
            quantity = Quantity(BigDecimal("500"), Unit.ML, 2),
        )
        val bd = calculatePrice(offer)
        assertEquals(BigDecimal("1000"), bd.normalizedQuantity)
        assertEquals("ml", bd.quantityUnit)
        assertEquals(BigDecimal("0.500000"), bd.cashUnitCost)
    }

    // ===================================================================
    // 複合シナリオ
    // ===================================================================

    @Test
    fun `drugstore scenario`() {
        val offer = Offer(
            id = "drugstore",
            productName = "洗剤",
            storeName = "ドラッグストア",
            price = BigDecimal("980"),
            taxIncluded = false,
            taxRate = BigDecimal("0.10"),
            quantity = Quantity(BigDecimal("500"), Unit.ML, 2),
            percentageDiscount = BigDecimal("0.10"),
            couponDiscount = BigDecimal("100"),
            pointRate = BigDecimal("0.01"),
        )
        val ctx = PurchaseContext(shippingFee = BigDecimal.ZERO)
        val bd = calculatePrice(offer, context = ctx)
        assertEquals(BigDecimal("870"), bd.payableNow)
        assertEquals(BigDecimal("8"), bd.earnedPoints)
        assertEquals(BigDecimal("862"), bd.effectiveCost)
    }

    @Test
    fun `online scenario`() {
        val offer = Offer(
            id = "online",
            productName = "洗剤",
            storeName = "通販",
            price = BigDecimal("330"),
            taxIncluded = true,
            taxRate = BigDecimal("0.10"),
            quantity = Quantity(BigDecimal("1000"), Unit.ML),
            pointRate = BigDecimal("0.05"),
        )
        val ctx = PurchaseContext(shippingFee = BigDecimal("80"))
        val bd = calculatePrice(offer, context = ctx)
        assertEquals(BigDecimal("410"), bd.payableNow)
        assertEquals(BigDecimal("20"), bd.earnedPoints)
        assertEquals(BigDecimal("390"), bd.effectiveCost)
    }
}

File: test/domain/price_comparison/pricing_calculator_test.dart
/// test/domain/price_comparison/pricing_calculator_test.dart
///
/// 税・割引・送料・ポイント・容量換算の計算順序を検証する。
///
/// 関連:
///   - lib/domain/price_comparison/pricing_calculator.dart
library;

import 'package:flutter_test/flutter_test.dart';
import 'package:hee_no_tane_app/domain/price_comparison/models.dart';
import 'package:hee_no_tane_app/domain/price_comparison/pricing_calculator.dart';

void main() {
  test('税から単位単価まで指定順で計算する', () {
    const offer = Offer(
      id: 'a',
      productName: '洗剤',
      storeName: 'A店',
      price: 1000,
      taxIncluded: false,
      taxRate: 0.10,
      percentageDiscount: 0.10,
      fixedDiscount: 50,
      couponDiscount: 40,
      pointRate: 0.01,
      shippingFee: 100,
      quantity: Quantity(
        value: 500,
        unit: Unit.gram,
        packageCount: 2,
        measureKind: MeasureKind.genericMass,
      ),
    );

    final result = calculatePrice(
      offer,
      context: const PurchaseContext(shippingFee: 100),
    );

    expect(result.taxAmount, 100);
    expect(result.basePriceInclTax, 1100);
    expect(result.discountAmount, 200);
    expect(result.payableNow, 1000);
    expect(result.earnedPoints, 10);
    expect(result.effectiveCost, 990);
    expect(result.normalizedQuantity, 1000);
    expect(result.cashUnitCost, 1);
    expect(result.effectiveUnitCost, closeTo(0.99, 0.000001));
  });

  test('税込398円の税額を36円として抽出する', () {
    const offer = Offer(
      id: 'a',
      productName: '食品',
      storeName: 'A店',
      price: 398,
      taxIncluded: true,
      taxRate: 0.10,
      earnedPoints: 0,
    );

    final result = calculatePrice(offer);

    expect(result.preTaxPrice, 362);
    expect(result.taxAmount, 36);
    expect(result.payableNow, 398);
    expect(result.effectiveCost, 398);
  });

  test('0.5円の境界をHALF_UPで1円に丸める', () {
    const offer = Offer(
      id: 'a',
      productName: '端数商品',
      storeName: 'A店',
      price: 5,
      taxIncluded: false,
      taxRate: 0.10,
      earnedPoints: 0,
    );

    final result = calculatePrice(offer);

    expect(result.taxAmount, 1);
    expect(result.payableNow, 6);
  });
}
